core/src/main/java/com/onelogin/saml2/http/HttpRequest.java

```diff
@@ -0,0 +1,129 @@
+package com.onelogin.saml2.http;
+
+import static com.onelogin.saml2.util.Preconditions.checkNotNull;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableMap;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Framework-agnostic representation of an HTTP request.
+ *
+ * @since 2.0.0
+ */
+public final class HttpRequest {
+    private final String requestURL;
+    private final Map<String, List<String>> parameters;
+
+    /**
+     * Creates a new HttpRequest.
+     *
+     * @param requestURL the request URL (up to but not including query parameters)
+     * @throws NullPointerException if requestURL is null
+     */
+    public HttpRequest(String requestURL) {
+        this(requestURL, Collections.<String, List<String>>emptyMap());
+    }
+
+    /**
+     * Creates a new HttpRequest.
+     *
+     * @param requestURL  the request URL (up to but not including query parameters)
+     * @param parameters the request query parameters
+     * @throws NullPointerException if any of the parameters is null
+     */
+    public HttpRequest(String requestURL, Map<String, List<String>> parameters) {
+        this.requestURL = checkNotNull(requestURL, "requestURL");
+        this.parameters = unmodifiableCopyOf(checkNotNull(parameters, "queryParams"));
+    }
+
+    /**
+     * @param name  the query parameter name
+     * @param value the query parameter value
+     * @return a new HttpRequest with the given query parameter added
+     * @throws NullPointerException if any of the parameters is null
+     */
+    public HttpRequest addParameter(String name, String value) {
```

Comment:
@luuuis  Instead of return a new HttpRequest, why we don't just modify the HttpRequest provided? Any reason for doing that? I think it will be more confuse.